### DISCUSSION TITLE
Getting all local process variables

### DISCUSSION THREAD

**Discussion by mimaom** [AMalininHere/camunda-worker-dotnet#50] [Q&A]
Hi Alexey

First of all - thank you for your great work on this project!

I have a question about getting local process variables. I would like to be able to get all local process varaibles from an external task, but without knowing their names. I have tried removing the array of process variable names from the HandlerVariables and just set the LocalVariables attribute to true, but still no process variables are returned. I have looked at the payload sent to Camunda in this case, and it looks somehting like this:

```
{
    "workerId": "Default",
    "maxTasks": 1,
    "usePriority": true,
    "asyncResponseTimeout": 10000,
    "topics": [
        {
            "topicName": "calculator",
            "lockDuration": 20000,
            "localVariables": false,
            "variables": [],
            "deserializeValues": false,
            "includeExtensionProperties": true
        },
        {
            "topicName": "longTask",
            "lockDuration": 60000,
            "localVariables": false,
            "variables": [],
            "deserializeValues": false,
            "includeExtensionProperties": true
        }
    ]
}
```

The "variables" attribute is passed as an empty array which is why Camunda returns nothing.

Looking at the Camunda REST API for the fetch and lock operation, it says:

![image](https://user-images.githubusercontent.com/16879883/158393493-25fdd5a4-55e5-4f08-bdd6-70140dfb630d.png)

If the varialbes attribute is not provided (missing or null) ALL varialbes will be returned. This behaviour combined with the option of only getting local varialbes will get me all the local variables of an external task - without knowing their names in advance - and that is  exactly what I am looking for.

What do you think about such a change?

Best regards
Michael

**Comment by TechnoBerry** [comment]
Hello Michael
Thank you for your feedback!

I think that behavior of fetching of all variables when handler is annotated with `[HandlerVariables(LocalVariables = true)]` or `[HandlerVariables]` (missing list of variable names) leads breaking of backward compatibility and may be dangerous for customers of the library.

The easiest way to support your case is adding ability to pass `null` as list of variable names. This way obliges you to explicitly and unambiguously say that you need all the variables. Of course, using of `null` may looks like a strange solution but it will work as expected.
Another simple to use and implement way is adding `AllVarialbes` boolean flag to `HandlerVariablesAttribute`. This flag will be used like that: `[HandlerVariables(AllVarialbes = true, LocalVariables = true)]` or `[HandlerVariables(AllVarialbes = true)]`

Best regards
Alexey

**Comment by mimaom** [comment]
Hello Alexey

Yes, I agree. I also think we should add a new flag to the HandlerVarialbes to avoid a breaking change. We will implement this and create a pull request for it.

Thanks.

Best regards
Michael

**Reply by TechnoBerry** [reply]
I will be glad to accept your pull request. Thank you!
